README.md
#Java2word#


The missing library to generate MS Word Documents from Java code without any special components.

_Coding for fun - it has to be enjoyable._


_* When I say no third part or external libraries, I mean there isn't any Microsoft library or special components. Please refer to "Dependencies" section below._

project has been migrated from: 
* https://code.google.com/p/java2word/
* http://java2word.blogspot.com.au/


##WHEN TO USE JAVA2WORD##

Suitable when you have that particular clients/customers/project where you want to export or create MS Word reports and make changes easily. Another situation is to generate "Release Notes" document as a part of the build.

If you need to create MS Word document that looks like a reports and data comes from a database, this is the perfect library for it.

You could use it in a web or non-web application.


##Just a quick taste of java2word##

```
IDocument myDoc = new Document2004();
myDoc.addEle(Heading1.with("Heading01").create());
myDoc.addEle(BreakLine.times(2).create()); //two break lines
myDoc.addEle(Paragraph.with("This document is an example of paragraph").create());

//then get the XML representation of the MS Word document
String myWord = myDoc.getContent(); 
```


##You can utilize the following elements:##

* Cover page with images or text
* Table of content based on Document Headings *
* Headings/Sessions/Title or Subtitles. Eg.: Heading1, Heading2
* Header with company's logo
* Footer with page numbers
* Paragraphs
* Page break
* Table with data coming from your database queries


**Believe me**: We first implemented this using Jasper Reports and the final result wasn't great. We created about 20 little reports, put all together as sub-reports, export them to RTF and opened it in MS Word. The RTF file had 5 MB and after saving as .DOC, file jumped to 40MB!

Jasper generates some dodge tables which are really annoying to edit their content.

It is so painful to implement this in Jasper. In addition to that, it requires a lot of manual work to put them together and you also have to manually convert to MS Word.


##WHEN TO NOT USE##

* Too fancy word documents with rich format/style.



##JAVA2WORD IS SIMPLE LIKETHIS##

Here is the code snippets:

```
 IDocument myDoc = new Document2004();
myDoc.addEle(Heading1.with("Heading01").create());
myDoc.addEle(Paragraph.with("This document is an example of paragraph").create());

myDoc.addEle(new Image("http://your_web_app/images/logo1w.png", ImageLocation.WEB_URL));


myDoc.addEle(new BreakLine(2)); //two break lines
```

Then:

```         
String myWord = myDoc.getContent();
```


Now you can add this to your ServletResponse:

```     
PrintWriter writer = servletResponse.getWriter();
writer.println(myWord);    
    
Then you will have your Microsoft Word document ready to be downloaded!

For non-web applications, you could create a local file:

File fileObj = new File("/home/leonardo/Desktop/Java2word_allInOne.doc");

PrintWriter writer = null;
try {
    writer = new PrintWriter(fileObj);
} catch (FileNotFoundException e) {
    e.printStackTrace();
}
String myWord = myDoc.getContent();

writer.println(myWord);
writer.close();            
    
```    

Take a look at the example projects to have a better understanding how to use it in a real world application.



##HOW TO GETTHE SERVLETRESPONSE REFERENCE##


###Using JBoss Seam###

If you are using JBoss Seam, you'll have something like (Full Working Class Example):


```
import java.io.PrintWriter;
import javax.faces.context.FacesContext;
import javax.servlet.http.HttpServletResponse;

import word.api.interfaces.IDocument;
import word.w2004.Document2004;
import word.w2004.elements.BreakLine;
import word.w2004.elements.Heading1;
import word.w2004.elements.Paragraph;
//probably other imports...


@Name("mySeamComponent")
public class reportCreator {

       public void getReport(){
                FacesContext fc = FacesContext.getCurrentInstance();
                
                HttpServletResponse servletResponse = (HttpServletResponse) fc
                                .getExternalContext().getResponse();

                //'UTF-8', 'ISO-8859-1' or nothing? up to you...
                //servletResponse.setContentType("application/msword; charset=UTF-8");
                servletResponse.setHeader("Content-disposition", "inline; filename="
                                + "myWordDoc.doc");
                
                PrintWriter writer = servletResponse.getWriter();


                //Create the word document
                IDocument myDoc = new Document2004();
                myDoc.addEle(Heading1.with("Heading01").create());
                myDoc.addEle(new BreakLine(2)); //two break lines
                myDoc.addEle(Paragraph.with("This document is an example of paragraph").create());

                String myWord = myDoc.getContent();


                writer.println(myWord);
                fc.responseComplete();
       }

}            
```



###Using Struts###

 public ActionForward execute(ActionMapping mapping, ActionForm form,
                HttpServletRequest request, HttpServletResponse response)
                throws Exception {



        //'UTF-8', 'ISO-8859-1' or nothing? up to you...
        //servletResponse.setContentType("application/msword; charset=UTF-8");
        response.setHeader("Content-disposition", "inline; filename="
                        + "myWordDoc.doc");
        
        PrintWriter writer = response.getWriter();

        writer.println(myWord);

        response.flushBuffer();

        return null;
}            
    
##DEPENDENCIES##

* xstream-1.3.1.jar : Only if you use Image component in your doc (needed for Base64 encoding). 

If you use maven, read the section "Java2Word with with Maven" below.



##WITH MAVEN##

At the moment, there is no Maven repository available for this jar. If you know any free place to publish the Jar file please let me know.

If you follow steps in "Set up your Development Environment" and successfully build the project, you should have in your local **~/.m2** folder some jar like: **~/.m2/repository/java2word/java2word/3.x/java2word-3.x.jar**.




##WITHOUT MAVEN?##

You just need to drop the the *java2word-X.X.jar* (obviously) into the lib folder of your web app. If you use Image in your doc you need *xstream-1.3.1.jar*.



##EXAMPLE PROJECTS##

Java project NOT using maven. Drop the jar (java2word-1.0.jar) and xstream-1.3.1.jar files in the lib directory. I utilized Struts in this example:


Take a look at the Struts Example project. There is a readme file there with steps for deployment.



##ALL-IN-ONE EXAMPLE - SEE EVERYTHING YOU CAN DO WITH JAVA2WORD.##

Please check out this page:
https://github.com/leonardoanalista/java2word/wiki/all-in-one-example


##SPECIAL CHARACTERS##

Please check out this page:
https://github.com/leonardoanalista/java2word/wiki/Encoding-Tips-and-Tricks
If there is a better way to do a replacement I am happy to hear about it.



##CREATING A NEW WORDELEMENT##

Lets imagine that you need some element that hasn't been implemented. Eg.: Bold text, Image, List...
Take a look at this page:
https://github.com/leonardoanalista/java2word/wiki/how-to-implement-a-new-element



##TEMPLATES WITHJAVA2WORD (BETA)##

I created a experimental solution for templates. Take a look at this page: 
https://github.com/leonardoanalista/java2word/wiki/templates



##I WANT TO HELPAND##
* build the project
* follow current pattern. Feel free to improve things. I wrote the code about 3 or 4 years ago, very late at night.
* write unit tests, check coverage, cyclomatic complexity, pmd and findbugs.
* test is in a real MS word document.
* raise a pull request




##JAVADOC##

I don't know how to host this in google code. The good thing is I commented in almost all methods. When you are writing your code, Eclipse (or whatever) should display the comment for you.

TODO: link to a component library reference.


####CURIOSITIES####

There was a real world project needed. However I really wanted to practice TDD skills and apply all code quality tools in java back in 2009. I decided to host the code at Google's SVN. I created the project and gave literally one line description just because field was required. After 5 days, one guy asked me about the documentation - I had to quickly add examples.

* I "forced" the usage some design patterns GOF (Strategy) just because I wanted to practice.
* I haven't used Windows since 2008.
* I use Mac OS and Linux and currently don't have MS Word installed. I use VM and Word Viewer on Wine' 
* I am now a front-end developer. I don't have a lot of time to alocate to this library. Feel free to send a pull request and contribute.  




_abraco a todos!_
_Leonardo Correa_




java2word/releaseNotes.txt
============= Version 3.3 - 07-May-2012  =============
    
    - issue 57 Facing problem in changing font of text.
    

============= Version 3.2 - 01-Feb-2012  =============


    - Experimental TableV2 (version 2) has been added. Possible to merge cells, format cells and all features of V1.
    - Issue 85: Create Image from InputStream
    
============= Version 3.1 - 01-Feb-2012  =============

Minor release. Applied patch for Identation and hiperlinks by whoahbot, Added some comments on methods, work in progress for special characters in Utils.java.

    - Issue 70: Created a method on Utils.class called replaceSpecialCharacters. This is under experiments.
    - Issue 82: Alternative for those who want to use paragraphs in table. 
              Now you can NOT pass null to a table cell, which doens't make sense anyway. You have to pass "".
      
    - Issue 46:   Adding indentation
    - Issue 45:   Adding hyperlinks to the document   
    - Removed all deprecated methods

Upgrade steps:
    - download this version or checkout the latest code and build your own jar file.

java2word/src/main/java/word/utils/ImageUtils.java
package word.utils;

import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

import javax.imageio.ImageIO;

import com.thoughtworks.xstream.core.util.Base64Encoder;

public class ImageUtils {

	public static String getImageHexaBase64(BufferedImage bufferedImage, String imageformat) {
		//System.out.println("@@@@@@@ IMAGE - getImageHexaBase64 NEW Way @@@@@@");
		ByteArrayOutputStream baos = new ByteArrayOutputStream(1000);
		//"gif"  "png" "jpeg"
		try {
			ImageIO.write(bufferedImage, imageformat.toString() , baos);
			baos.flush();
		} catch (IOException e) {
			e.printStackTrace();
			throw new RuntimeException("Erro generating Base64 for the image", e);
		}
		byte[] resultImageAsRawBytes = baos.toByteArray();

		String encodedString = new Base64Encoder().encode(resultImageAsRawBytes);
		//String encodedString = Base64.encodeBytes(resultImageAsRawBytes);
		return encodedString;
	}


/*
	private final static String XML_DOC_HEADER = "<?xml version=\"1.0\" ?>\n";
	public static final String[] hexLookupTable = { "00", "01", "02", "03",
			"04", "05", "06", "07", "08", "09", "0a", "0b", "0c", "0d", "0e",
			"0f", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19",
			"1a", "1b", "1c", "1d", "1e", "1f", "20", "21", "22", "23", "24",
			"25", "26", "27", "28", "29", "2a", "2b", "2c", "2d", "2e", "2f",
			"30", "31", "32", "33", "34", "35", "36", "37", "38", "39", "3a",
			"3b", "3c", "3d", "3e", "3f", "40", "41", "42", "43", "44", "45",
			"46", "47", "48", "49", "4a", "4b", "4c", "4d", "4e", "4f", "50",
			"51", "52", "53", "54", "55", "56", "57", "58", "59", "5a", "5b",
			"5c", "5d", "5e", "5f", "60", "61", "62", "63", "64", "65", "66",
			"67", "68", "69", "6a", "6b", "6c", "6d", "6e", "6f", "70", "71",
			"72", "73", "74", "75", "76", "77", "78", "79", "7a", "7b", "7c",
			"7d", "7e", "7f", "80", "81", "82", "83", "84", "85", "86", "87",
			"88", "89", "8a", "8b", "8c", "8d", "8e", "8f", "90", "91", "92",
			"93", "94", "95", "96", "97", "98", "99", "9a", "9b", "9c", "9d",
			"9e", "9f", "a0", "a1", "a2", "a3", "a4", "a5", "a6", "a7", "a8",
			"a9", "aa", "ab", "ac", "ad", "ae", "af", "b0", "b1", "b2", "b3",
			"b4", "b5", "b6", "b7", "b8", "b9", "ba", "bb", "bc", "bd", "be",
			"bf", "c0", "c1", "c2", "c3", "c4", "c5", "c6", "c7", "c8", "c9",
			"ca", "cb", "cc", "cd", "ce", "cf", "d0", "d1", "d2", "d3", "d4",
			"d5", "d6", "d7", "d8", "d9", "da", "db", "dc", "dd", "de", "df",
			"e0", "e1", "e2", "e3", "e4", "e5", "e6", "e7", "e8", "e9", "ea",
			"eb", "ec", "ed", "ee", "ef", "f0", "f1", "f2", "f3", "f4", "f5",
			"f6", "f7", "f8", "f9", "fa", "fb", "fc", "fd", "fe", "ff" };


	public void emptyDocTest() {
		Document2000 myDoc = new Document2000();
		// System.out.println(myDoc.getContent());

		try {
			// BufferedImage img = ImageIO.read(new File(new
			// URI("http://localhost:8080/app02/img/dtpick.gif") ));
			// OK!!! BufferedImage img = ImageIO.read(new
			// File("/Users/leonardo_correa/Desktop/icons_corrup/robber-thumb1893761.jpg"));
			// BufferedImage img = ImageIO.read(new
			// File("http://localhost:8080/app02/img/dtpick.gif"));

			URL url = new URL("http://localhost:8080/app02/img/dtpick.gif");
			BufferedImage img = ImageIO.read(url);
			System.out.println("H: " + img.getHeight());
			System.out.println("W: " + img.getWidth());
			System.out.println(img.toString());
			System.out.println("########################################");

			// createBase64XMLDoc("/Users/leonardo_correa/Desktop/icons_corrup/quote.gif",
			// "xxxx.txt");

			// ##### Goood code to get the hexa from a picture in the SAME
			// server ###
			File file = new File(
					"/Users/leonardo_correa/Desktop/icons_corrup/quote.gif");
			BufferedInputStream bis = new BufferedInputStream(
					new FileInputStream(file));
			int bytes = (int) file.length();
			byte[] buffer1 = new byte[bytes];
			int readBytes1 = bis.read(buffer1);
			bis.close();

			String encodedString = Base64.encode(buffer1);
			System.out.println(encodedString);
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

	}

	public String padHexString(String hexNum) {
		if (hexNum.length() < 2) {
			return "0" + hexNum;
		}
		return hexNum;
	}

	public void createBase64XMLDoc(String filename, String outFilename) {
		try {
			byte[] buffer = readFile(filename);
			int readBytes = buffer.length;

			long startTime = (new java.util.Date()).getTime();
			StringBuffer strXMLDoc = new StringBuffer(XML_DOC_HEADER);
			strXMLDoc.append("<image name=\"" + filename + "\">");
			String encodedString = Base64.encode(buffer);
			strXMLDoc.append(encodedString);
			strXMLDoc.append("</image>");
			long endTime = (new java.util.Date()).getTime();
			long diffTime = endTime - startTime;
			System.out.println("Total conversion time for " + readBytes
					+ " was " + diffTime + " msec or " + ((double) readBytes)
					/ ((double) diffTime) + " bytes/msec.");

			writeFile(outFilename, strXMLDoc.toString());
		} catch (IOException e) {
			System.out.println("Our exception is: " + e);
		}
	}

	public void createHexXMLDoc(String filename, String outFilename) {
		try {
			byte[] buffer = readFile(filename);
			long startTime = (new java.util.Date()).getTime();
			int readBytes = buffer.length;
			StringBuffer strXMLDoc = new StringBuffer(XML_DOC_HEADER);
			strXMLDoc.append("<image name=\"" + filename + "\">");
			for (int i = 0; i < readBytes; i++) {
				strXMLDoc.append(hexLookupTable[0xff & buffer[i]]);
			}
			strXMLDoc.append("</image>");
			long endTime = (new java.util.Date()).getTime();
			long diffTime = endTime - startTime;
			System.out.println("Total conversion time for " + readBytes
					+ " was " + diffTime + " msec or " + ((double) readBytes)
					/ ((double) diffTime) + " bytes/msec.");

			writeFile(outFilename, strXMLDoc.toString());
		} catch (IOException e) {
			System.out.println("Our exception is: " + e);
		}
	}

	public byte[] readFile(String filename) throws IOException {
		File file = new File(filename);
		BufferedInputStream bis = new BufferedInputStream(new FileInputStream(
				file));
		int bytes = (int) file.length();
		byte[] buffer = new byte[bytes];
		int readBytes = bis.read(buffer);
		bis.close();
		System.out.println("Read " + readBytes + " and expected to read "
				+ bytes);
		return buffer;
	}

	public void writeFile(String filename, String data) throws IOException {
		File file = new File(filename);
		BufferedOutputStream bos = new BufferedOutputStream(
				new FileOutputStream(file));
		bos.write(data.getBytes());
		bos.close();

		System.out.println("Wrote " + data.getBytes().length + " to file "
				+ filename);
		return;
	}

*/

}
